### PR TITLE
parallelize x.py test tidy

### PR DESCRIPTION
**Pull request by the8472**
Running tidy on individual commits when rewriting git history was somewhat of an annoyance, so I have parallelized it a bit.

running `time ./x.py test tidy` with warm IO caches:

old:

```
real	0m11.123s
user	0m14.495s
sys	0m5.227s
```

new:

```
real	0m1.834s
user	0m13.545s
sys	0m3.094s
```

There's further room for improvement (<0.9s should be feasible) but that would require bigger changes.

### CODE REVIEW

File: src/bootstrap/format.rs [hunk 1]
```diff
@@ -101,19 +107,58 @@ pub fn format(build: &Build, check: bool) {
...
+        .config
+        .initial_rustfmt
+        .as_ref()
+        .unwrap_or_else(|| {
+            eprintln!("./x.py fmt is not supported on this channel");
+            std::process::exit(1);
+        })
+        .to_path_buf();
+    let src = build.src.clone();
+    let (tx, rx): (SyncSender<PathBuf>, _) = std::sync::mpsc::sync_channel(128);
```
Comment (pickfire) [inline src/bootstrap/format.rs:120]: Why 128?

File: src/bootstrap/format.rs [hunk 2]
```diff
@@ -101,19 +107,58 @@ pub fn format(build: &Build, check: bool) {
...
+    let max_processes = num_cpus::get() * 2;
+
+    // spawn child processes on a separate thread so we can batch entries we have received from ignore
+    let thread = std::thread::spawn(move || {
+        let mut children = VecDeque::new();
+        while let Ok(path) = rx.recv() {
+            // try getting a few more paths from the channel to amortize the overhead of spawning processes
+            let paths: Vec<_> = rx.try_iter().take(7).chain(std::iter::once(path)).collect();
+
+            let child = rustfmt(&src, &rustfmt_path, paths.as_slice(), check);
```
Comment (pickfire) [inline src/bootstrap/format.rs:135]: Should we keep spawning processes? Why not just have all the paths and distribute equally among threads? I mean some files could be finished faster than others but will it matter a lot compared to keep spawning new processes?

File: src/bootstrap/format.rs (hunk 1 above)
Comment (the8472) [reply src/bootstrap/format.rs:120]: No strong reason, I just avoid unbounded queues by default and a bounded one requires me to choose some capacity. 128 items are enough to immediately spawn 16 `rustfmt` processes assuming the queue is full. The loop below picks up to 8 items per iteration. If you have more than 16 hardware threads then spawning additional ones will take a few moments for `ignore::Walk` to refill the queue, which should be fine most of the time.

File: src/bootstrap/format.rs (hunk 2 above)
Comment (the8472) [reply src/bootstrap/format.rs:135]: That would lead to better resource utilization, yes. But since the `rustfmt` CLI doesn't have a way to feed it additional files while it's running that would require including it as a library instead.

I haven't looked into why tidy calls the CLI instead of using it as lib. I assume there's a reason for that so I went for a smaller change.

File: src/bootstrap/format.rs (hunk 1 above)
Comment (pickfire) [reply src/bootstrap/format.rs:120]: Should we calculate this automatically with `num_cpus * 8`?

Comment (the8472) [reply src/bootstrap/format.rs:120]: I don't think it's necessary. Traversing directories should keep this filled faster than starting and _waiting_ for processes.

**Comment by the8472** [comment]
rebased and switched from rayon to crossbeam-utils.

File: src/bootstrap/format.rs [hunk 3]
```diff
@@ -101,19 +107,58 @@ pub fn format(build: &Build, check: bool) {
...
+    // there is a lot of blocking involved in spawning a child process and reading files to format.
+    // spawn more processes than available cores to keep the CPU busy
+    let max_processes = num_cpus::get() * 2;
+
+    // spawn child processes on a separate thread so we can batch entries we have received from ignore
+    let thread = std::thread::spawn(move || {
+        let mut children = VecDeque::new();
+        while let Ok(path) = rx.recv() {
+            // try getting a few more paths from the channel to amortize the overhead of spawning processes
+            let paths: Vec<_> = rx.try_iter().take(7).chain(std::iter::once(path)).collect();
```
Comment (jyn514) [inline src/bootstrap/format.rs:133]: Have you tried moving this number up or down to see if it affects the time?

File: src/bootstrap/format.rs [hunk 4]
```diff
@@ -101,19 +107,58 @@ pub fn format(build: &Build, check: bool) {
...
             let entry = t!(entry);
             if entry.file_type().map_or(false, |t| t.is_file()) {
-                rustfmt(src, &rustfmt_path, &entry.path(), check);
+                t!(tx.send(entry.into_path()));
             }
             ignore::WalkState::Continue
         })
     });
+
+    drop(tx);
```
Comment (jyn514) [inline src/bootstrap/format.rs:161]: Why do you explicitly drop `tx` here?

Comment (the8472) [reply src/bootstrap/format.rs:161]: to close the channel, otherwise the thread would wait for new messages forever and not terminate.

File: src/bootstrap/format.rs (hunk 3 above)
Comment (the8472) [reply src/bootstrap/format.rs:133]: 0: 0m2.182s
7: 0m1.742s
15: 0m1.679s
31: 0m1.701s

File: src/tools/tidy/src/bins.rs [hunk 5]
```diff
@@ -32,9 +32,12 @@ pub fn check(path: &Path, output: &Path, bad: &mut bool) {
     // readily create a file there to test.
     //
     // See #36706 and #74753 for context.
-    let mut temp_path = path.join("tidy-test-file");
+    //
+    // We also append the thread ID to avoid threads trampling on each others files.
+    let file_name = format!("t{}.tidy-test-file", std::thread::current().id().as_u64());
```
Comment (Mark-Simulacrum) [inline src/tools/tidy/src/bins.rs:37]: This implies we're visiting the same path more than once, which seems surprising to me - why is this necessary?

**Review by Mark-Simulacrum — COMMENTED** [review]
The ~5x win you list is nice, and it doesn't look like this adds too much complexity, so it's overall not too bad... I am inclined to merge, but would like to spend a bit of time thinking about it.

File: src/tools/tidy/src/bins.rs (hunk 5 above)
Comment (the8472) [reply src/tools/tidy/src/bins.rs:37]: Maybe this isn't actually needed. I did this change before the one in `filter_dirs`. I'll retest with only the `filter_dirs` change.

File: src/bootstrap/format.rs (hunk 4 above)
Comment (jyn514) [reply src/bootstrap/format.rs:161]: Does that mean if an error occurs `tidy` will hang forever, because tx gets dropped last?

Comment (the8472) [reply src/bootstrap/format.rs:161]: Not sure what scenario you have in mind. But if walkdir panics then the program will terminate anyway since it's on the main thread. If the `rustfmt`-spawning thread panics then `rx` gets dropped which should result in `tx` encountering an error and terminating the walk.

**Comment by Mark-Simulacrum** [comment]
Also, one thing we'll want to do is respect -j when passed to x.py, for both tidy and fmt checking. Ideally we'd also utilize the jobserver crate to properly handle parallelism when running under a larger make, but that seems complicated.

**Comment by the8472** [comment]
> Ideally we'd also utilize the jobserver crate to properly handle parallelism when running under a larger make, but that seems complicated.

Bootstrap currently doesn't provide a jobserver, right? Only cargo does afaict.

Getting the testsuite to run under a jobserver would be nice since it could unlock some additional room for parallelism - e.g. building stage 0 compiletest or rustdoc concurrent with the stage 1 compile - but that seems like a much bigger task overall.

File: src/bootstrap/format.rs (hunk 4 above)
Comment (pickfire) [reply src/bootstrap/format.rs:161]: @jyn514 It is a pattern when using channels such since if it does not explicitly dropped (or wait for it to goes out of scope), I see this pattern quite often. I wanted to comment if we should put this as a comment to show those does never seen this but I didn't? Maybe we should drop a comment saying that we need to drop the last transfer reference count so that the receiving part will stop waiting?

By the way, happy Chinese New Year!

Comment (the8472) [reply src/bootstrap/format.rs:161]: I think such an explanation/example would be more appropriate somewhere in the `std::sync` module if it's not there already.

**Comment by the8472** [comment]
`-j1` makes things slow again.

@rustbot label: +S-waiting-on-review -S-waiting-on-author

File: src/tools/tidy/src/bins.rs (hunk 5 above)
Comment (the8472) [reply src/tools/tidy/src/bins.rs]: > thread '<unnamed>' panicked at 'Deleted temp file: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/tidy/src/bins.rs:43:46

Looks like adding the thread ID was necessary after all.

Comment (Mark-Simulacrum) [reply src/tools/tidy/src/bins.rs:37]: I'm not satisfied with just an error message - something is wrong with at least my mental model of the code, and that leads me to believe this may just be hiding some other bug. I would like to see an explanation for why the error is happening.

Comment (the8472) [reply src/tools/tidy/src/bins.rs:37]: One possibility is that the first create fails and then the 2nd attempt uses the output dir which is shared between the passes.

https://github.com/rust-lang/rust/blob/3c10a880eca60379343a6c6d19dd5bda38ead55d/src/tools/tidy/src/bins.rs#L36-L38

Comment (Mark-Simulacrum) [reply src/tools/tidy/src/bins.rs:37]: Can we not parallelize this pass perhaps to avoid this? I'm not convinced by your explanation I guess and it feels subtle; I'd rather have slow correct code than wrong parallel code :)

**Comment by Mark-Simulacrum** [comment]
Local measurement suggests that 70% of the tidy runtime comes from rustfmt, so it might be good to just skip the parallelization of the tidy part (avoiding complexity there), and retain just the rustfmt parallelism; ignore is already walking the tree in parallel today so there's not much added complexity there.

Does that make sense to you?

File: src/tools/tidy/src/bins.rs (hunk 5 above)
Comment (the8472) [reply src/tools/tidy/src/bins.rs:37]: We could, but I don't quite understand the concern. The tests are largely non-mutating. This is the only mutating part and the fix I did would be *necessary* for correctness anyway due to to the possibility mentioned above.

So what do you fear might be going wrong beyond this?

**Comment by the8472** [comment]
I'd prefer if the runtime were even lower, rather than higher. But if that's the only way to get this merged... what choice do I have?

I would prefer to understand your concerns though.

File: src/tools/tidy/src/bins.rs (hunk 5 above)
Comment (Mark-Simulacrum) [reply src/tools/tidy/src/bins.rs:37]: The modified comment in the code doesn't suggest a correctness fix to me, so maybe I overlooked that - if there's a correctness fix here, I'm not yet seeing how we've avoided the problem by way of thread ID... does this problem occur before your PR? If so, then I don't see how adding thread ID helps, as previously it'd always be zero, right?

Comment (the8472) [reply src/tools/tidy/src/bins.rs:37]: What I was trying to say is that it is a *necessary* part of making these checks concurrent because the code path creating the files may create them in either the source or the output folder and the output folder is shared. So if two checks run concurrently then they might create the file twice and try to delete it twice which will fail if it's not a unique file name.

Your concern seems to be that it's not *sufficient*, but I don't see anything else that might trigger this.

**Comment by the8472** [comment]
I'll split the tidy changes into a separate PR so they can be discussed on their own.

**Comment by Mark-Simulacrum** [comment]
Well, there is a cost to the parallelism being added to tidy - crossbeam is not the lightest of dependencies, and I'd expect a bug report down the line that we're not handling jobserver tokens right (when -j1 is passed to make, rather than x.py directly).

Ultimately I think we can agree that there's a cost to the added parallelism; reduced but not eliminated by the safety guarantees Rust provides. I'm not sure that justifying that tradeoff for a few seconds of time is worth it, but it's also true that I don't really run tidy locally at all; others' workflows may involve calling it more often, and that may mean that they care more about this. It's hard for me to accept added code and complexity for something that at least for me is primarily a CI cost.

The rustfmt parallelism is presumably where the majority of the wins this PR brings come from, and so I'm more willing to accept those - they're also in theory less impactful from a maintenance perspective, as they're entirely local (i.e., we just call rustfmt with more processes at once). I wouldn't be surprised if we could get similar wins without any parallelism, just batching ~15-20 files to rustfmt at a time, but maybe my calculus is off there. Finding a way to avoid --skip-children might be even more rewarding

**Comment by the8472** [comment]
> others' workflows may involve calling it more often, and that may mean that they care more about this.

Yeah, I run it frequently when doing interactive git rebase. I was also thinking of using it as a commit hook but it's still too slow for that.

> I wouldn't be surprised if we could get similar wins without any parallelism, just batching ~15-20 files to rustfmt at a time

It's already batching (N = 8). I measured performance win from batch sizes in [this discussion](https://github.com/rust-lang/rust/pull/81833#discussion_r575095201) and larger batches provide barely any speedups. I also ran this under perf, most of the wins are from parallelism. 

> Well, there is a cost to the parallelism being added to tidy - crossbeam is not the lightest of dependencies

I can write this without crossbeam if you want, 1 thread per check, up to `-j` threads or something. Anyway... we can discuss it in a separate PR soon. 

> and I'd expect a bug report down the line that we're not handling jobserver tokens right (when -j1 is passed to make, rather than x.py directly).

Eh, this only runs for a few seconds, so it shouldn't screw with anyone's build system. I mean jobserver support for all of bootstrap is a great idea, but by no means trivial, at least if we want to fully utilize it e.g. by parallelizing build phases and ordering them before test phases. But I don't see it as a killer feature in this particular case.

There are much bigger fish not obeying jobserver tokens, e.g. compiletest which runs for minutes.

**Comment by the8472** [comment]
Updated timings with just the rustfmt changes

```
real	0m2.847s
user	0m13.466s
sys	0m1.782s
```

File: src/tools/tidy/src/bins.rs (hunk 5 above)
Comment (Mark-Simulacrum) [reply src/tools/tidy/src/bins.rs]: FWIW (I guess we're moving it to a separate PR) I think I'd suggest we move this out of the parallel code or cache the results; they shouldn't change and doing it just once seems simpler than trying to make sure we don't clobber each other. It also helps us avoid accidentally getting different results in different threads, which could be quite confusing.

**Comment by Mark-Simulacrum** [comment]
@bors r+ rollup

Ok, thanks for splitting things out. tidy and rustfmt are pretty orthogonal anyway so it seems good to land them separately. I think I'm still interested in seeing the tidy changes - see my last comment on the one direct problem I wanted to remove before landing those changes.

File: src/bootstrap/format.rs (hunk 4 above)
Comment (pnkfelix) [reply src/bootstrap/format.rs:161]: Looks like @pickfire put up PR #80269 to serve as that requested explanation.
